### PR TITLE
df.dfpde faster execution

### PR DESCRIPTION
**Pull request by ilario**
The `df.dfpde` function runs millions of times in each simulation (e.g. for running the example in the readme, including `equilibrate` and `doJV`, it runs 3.3 million times).
Currently it accounts for half of the total simulation time.
These changes tries to increase its efficiency without making the code too much ugly.
Even faster code could be obtained moving all the `if` checks outside of the `df.dfpde` function and writing various versions of it including 0, 1 or 2 ions. But this would mean much uglier code.
In the current set of changes the less nice part is the repetition of the `S_potential` expression.
And the slowest part of the function are the two `find` lines, I tried various alternatives (like ismembc2 and builtin _ismemberhelper) with no luck.
I am not super sure that all and each of the commits improve the speed, but must do.

Before the changes, the run time of the `equilibrate` example from the readme was 25.3 s (on my laptop) and after the changes is 17.9 s (further decreases to 17.7 s after applying also #38; further decreases to 13.8 s modifying pdepe for passing over the positional index, see [this question](https://uk.mathworks.com/matlabcentral/answers/518716-how-to-get-positional-index-from-pdepe-into-pde)).
Before the changes, the run time of the `doJV` example from the readme was 31.2 s and after the changes is 19.4 s (further decreases to 18.0 s after applying also #38; further decreases to 14.5 s modifying pdepe for passing over the positional index, see [this question](https://uk.mathworks.com/matlabcentral/answers/518716-how-to-get-positional-index-from-pdepe-into-pde)).
Using Matlab's profile function over the equilibrate and the doJV commands, the execution time of df.dfpde gets reduced to the half.

### CODE REVIEW

**Comment by ilario** [comment]
TL;DR: no improvement in further tests

I tested a couple more alternatives, in which the if checks are outside the `dfpde` function.
Instead of writing many different versions of `dfpde` function, I tried to generate it from a string and the string generated depending on the number of mobile ions and the generation dependence with time.
The problem is that `str2func` can generate only anonymous functions.

In this example:  https://github.com/ilario/Driftfusion/blob/dfpde3/Core/df.m I generated a function for returning F and S values and this function was used by `dfpde` function. The time needed for calling the new function compensates the removal of the if checks. In order to use only one function for generating both F and S I used the `deal` function, but the additional time needed by the `deal` function compensated the difference between calling two or one function.

In this other example: https://github.com/ilario/Driftfusion/blob/dfpde2/Core/df.m in order to avoid to call more than one function, the whole `dfpde` was generated using `str2func`. The problem is that the `i` index cannot be calculated inside of it, so the `pdepe.m` Matlab's function has to be modified to pass the index to the `dfpde` function as an argument.

Neither of the two modifications resulted in a considerable improvement of performances (over the code contained in this pull request) but they made the code much more dirty, so the current PR is still my favourite option.

**Comment by barnesgroupICL** [comment]
I accidentally accepted all of the commits to df.dfpde faster execution #39 when I was intending going through each commit individually. Now it the pull request has moved into ‘closed’ status. I’m afraid I can’t easily see how to reopen it… maybe it’s possible for you to do it?

In general I think we need to find a balance here between efficiency of the code and readability hence not accepting all the suggested changes without the appropriate scrutiny.

**Comment by ilario** [comment]
> Does this genuinely speed things up? Looks like you still have to use find(x_ihalf <=x) but then have an additional calculation to perform?

I suppose you refer to this:

https://github.com/ilario/Driftfusion/blob/1fdbf23a4e9c5db4ec52c66178102738b93fdec4/Core/df.m#L220-L223

The find is usually the most time consuming step, you can check that with `profile on; df(run_some_simulation); profile viewer`.

To perform 2 small find is a bit faster than a large find (I tested it when making the change, I can repeat the test).

What would be much faster would be to have the index directly from pdepe (see https://uk.mathworks.com/matlabcentral/answers/518716-how-to-get-positional-index-from-pdepe-into-pde ) but this would mean to distribute a modified copy of pdepe (which likely we cannot do due to Matlab's license?).

> This makes a lot of sense. Is it more efficient to use 'otherwise' in the switch rather than case 0? I think case 0 is probably safer for future expansion?

I suppose this refers to:

https://github.com/ilario/Driftfusion/blob/1fdbf23a4e9c5db4ec52c66178102738b93fdec4/Core/df.m#L197

and it has no influence on the performance, as it is out of the dfpde function, so if you think that having case 0 is better, that's perfect for me.

> Now it the pull request has moved into ‘closed’ status. I’m afraid I can’t easily see how to reopen it… maybe it’s possible for you to do it?

I'll make a new PR including also the aforementioned `case 0` change.
